### PR TITLE
Document how to use yapf

### PR DESCRIPTION
**Pull request by alok**
This is a cleaner version of #2158. @ericl Please take a look at this.

### CODE REVIEW

File: doc/source/development.rst [hunk 1]
```diff
@@ -54,7 +54,15 @@ helpful.
    something like ``flake8 ray/python/ray/worker.py``. You may need to first run
    ``pip install flake8``.
 
-5. **Inspecting Redis shards by hand:** To inspect the primary Redis shard by
+5. **Autoformatting code**. We use ``yapf``
+   https://github.com/google/yapf for linting, and the config file is
+   located at ``.style.yapf``. We recommend adding ``.travis/yapf.sh``
+   to git's ``pre-push`` hook, so that you never have to worry about
```
Comment (ericl) [inline doc/source/development.rst]: I'm curious why this is so slow, we only have ~90k lines of Python (even including rllib / dataframes), so it seems weird that it takes almost a minute to format that. I don't think we can recommend this unless we get formatting time down to some thing more reasonable, e.g. 5-10 seconds.

File: doc/source/development.rst [hunk 2]
```diff
@@ -54,7 +54,15 @@ helpful.
...
+   your code formatting ever again. Note that RLlib is not currently
+   autoformatted, though running ``yapf`` over the individual files you
+   edit with ``yapf -i -p -r <file1> <file2> <...>`` is a good idea.
```
Comment (ericl) [inline doc/source/development.rst]: How about this instead?
```
yapf -i -p `git diff --name-only upstream/master | grep .py`
```

File: .travis/yapf.sh [hunk 3]
```diff
@@ -9,14 +9,33 @@ builtin cd "$(dirname "${BASH_SOURCE:-$0}")"
...
+        --exclude 'python/ray/cloudpickle' \
+        --exclude 'python/ray/dataframe' \
+        --exclude 'python/ray/rllib' \
+        -- \
+        test python
+}
+
+# This flag lints individual files. --files *must* be the first command line
+# arg to use this option.
+if [[ "$1" == '--files' ]]; then
```
Comment (ericl) [inline .travis/yapf.sh]: Why this change? --files isn't used below.

File: doc/source/development.rst (hunk 1 above)
Comment (alok) [reply doc/source/development.rst]: Takes about 20 seconds on my MacBook

Comment (alok) [reply doc/source/development.rst]: Even a minute isn't that bad for pre push since you push far less than you commit

Comment (ericl) [reply doc/source/development.rst]: It must depend on your hardware then, on my laptop it takes twice a long (4x i7 cores). Still, 20s is an eternity when pushing typically takes just a couple seconds.

File: .travis/yapf.sh (hunk 3 above)
Comment (alok) [reply .travis/yapf.sh]: This is a poor man's argparse since I hate messing with bash's getopts. If `--files` is the first argument, then only the files passed are formatted.

File: doc/source/development.rst (hunk 1 above)
Comment (alok) [reply doc/source/development.rst]: The only one that's faster is [`black`](https://github.com/ambv/black), since it checks if the files on disk have been changed. 

I would be perfectly happy to switch to it, since it offers no configuration at all, which makes debates about style moot, except I doubt the rest of the team would, given the lack of said configuration.

File: doc/source/development.rst (hunk 2 above)
Comment (robertnishihara) [reply doc/source/development.rst]: Not sure it's actually a good idea to run `yapf` on the files you edit because that will introduce a lot of changes in places you didn't mean to change and will make it hard to see what your PR is doing.

File: doc/source/development.rst (hunk 1 above)
Comment (robertnishihara) [reply doc/source/development.rst]: FWIW, I've always been doing `git status` and then passing all changed files into

```
yapf --style=.style.yapf -i [all changed files]
```

File: doc/source/development.rst (hunk 2 above)
Comment (ericl) [reply doc/source/development.rst]: Hm I see, perhaps the --exclude flags would work here? Though then the cmd line becomes a bit long so you probably want to check in the script.

It would be great to have a ./yapf script that formats all relevant files and also is fast enough to run all the time.

Comment (alok) [reply doc/source/development.rst]: This problem is actually why I've been pushing so annoyingly hard for autoformat everywhere, all the time. If we bit the bullet, formatted all the python code and read the ugly PR, and then put a pre-*commit* hook (we could copy [this script from go](https://tip.golang.org/misc/git/pre-commit)), no change in the codebase would ever introduce line noise, because it would always be formatted the same. Since we'd only ever format changed files, it'd be fast enough to be unnoticeable too. It could later open up the possibility of [mechanical source transformation](https://blog.golang.org/go-fmt-your-code).

A pre-push hook is a halfway measure, but is still far better than what's currently done. It also saves having to write overly complex scripts for what's really a problem with standards. 

30k lines of ray were autoformatted without any major issues. I think the rest of the python code could be done too. I'd be happy to review, since code formatting hardly requires deep knowledge of the actual code being run.

Any PRs against `master` could run yapf in their final commit and then only the diff would appear in the github PR review tool, so it wouldn't be a pain to review.

**Comment by alok** [comment]
@ericl Can you see if the latest commit sped it up at all? Try `.travis/yapf.sh --all`.

File: doc/source/development.rst (hunk 2 above)
Comment (ericl) [reply doc/source/development.rst]: You can still add a pre-commit hook, just with --excludes for now, no?

FWIW I think the full linting will happen, it's just that we need to find some quiescent time for the sub-projects and then someone can do a quick 3am push to reformat everything.

**Comment by ericl** [comment]
It's still 25 seconds. Why not combine the changed file detection and --all though? That makes it instantaneous for most PRs.

Btw you should add a symlink from ~/.travis/yapf.sh to ~/yapf.sh, since the .travis files are hidden for most developers.

**Comment by alok** [comment]
I guess it makes sense to format just the files that differ from `master`,
since we almost always end up merging into there anyway. That's few enough
files to be instant, so it does double duty as a pre-push and pre-commit
hook.

**Comment by alok** [comment]
@ericl The latest script should do what you want, can you test it out?

File: .travis/yapf.sh [hunk 4]
```diff
@@ -9,15 +9,24 @@ builtin cd "$(dirname "${BASH_SOURCE:-$0}")"
...
         "$@"
 }
 
-lint_all() {
+format_changed() {
+    # Formats python files that changed in last commit
+    git diff --name-only HEAD~1 HEAD | grep '\.py$' | xargs -P 5 \
+        yapf \
+        --style "$ROOT/.style.yapf" \
+        --in-place --recursive --parallel
```
Comment (ericl) [inline .travis/yapf.sh]: missing --exclude

File: .travis/yapf.sh [hunk 5]
```diff
@@ -9,15 +9,24 @@ builtin cd "$(dirname "${BASH_SOURCE:-$0}")"
...
         --style "$ROOT/.style.yapf" \
         --in-place --recursive --parallel \
         -- \
         "$@"
 }
 
-lint_all() {
+format_changed() {
+    # Formats python files that changed in last commit
+    git diff --name-only HEAD~1 HEAD | grep '\.py$' | xargs -P 5 \
```
Comment (ericl) [inline .travis/yapf.sh]: This isn't quite right. Usually you want to diff from the last `master` commit, not the one previous.

File: .travis/yapf.sh (hunk 4 above)
Comment (alok) [reply .travis/yapf.sh]: Oops, didn't push the actual latest version, that's changed. I'd really rather avoid adding an --excludes flag. If someone wants to lint specific files/dirs, they can run `yapf.sh --files <files or dirs>`.

File: doc/source/development.rst [hunk 6]
```diff
@@ -54,13 +54,15 @@ helpful.
...
    ``pip install flake8``.
 
-5. **Autoformatting code**. We use ``yapf``
-   https://github.com/google/yapf for linting, and the config file is
-   located at ``.style.yapf``. We recommend adding ``.travis/yapf.sh``
-   to git's ``pre-push`` hook, so that you never have to worry about
-   your code formatting ever again. Note that RLlib is not currently
-   autoformatted, though running ``yapf`` over the individual files you
-   edit with ``.travis/yapf.sh --files <file1> <file2> <...>`` is a good idea.
+5. **Autoformatting code**. We use ``yapf`` https://github.com/google/yapf for
```
Comment (ericl) [inline doc/source/development.rst]:
```
5. **Autoformatting code**. We use ``yapf`` https://github.com/google/yapf for
   linting, and the config file is located at ``.style.yapf``. We recommend
   running ``.travis/yapf.sh`` prior to pushing to format changed files.
   Note that some projects such as dataframes and rllib are currently excluded.
```

Can you make the above workflow work?

Comment (alok) [reply doc/source/development.rst]: It should with the latest change.

**Review by ericl — COMMENTED** [review]
I pushed some suggested changes. Let me know if I'm missing anything.

File: .travis/yapf.sh [hunk 7]
```diff
@@ -9,17 +9,49 @@ builtin cd "$(dirname "${BASH_SOURCE:-$0}")"
...
+
+YAPF_EXCLUDES="--exclude 'python/ray/cloudpickle' \
+--exclude 'python/ray/dataframe' \
+--exclude 'python/ray/rllib' \
+--exclude 'python/build' \
+--exclude 'python/ray/pyarrow_files' \
+--exclude 'python/ray/core/src/ray/gcs' \
+--exclude 'python/ray/common/thirdparty'"
+
+UPSTREAM_MASTER=${RAY_UPSTREAM_BRANCH:-upstream/master}
```
Comment (alok) [inline .travis/yapf.sh]: This doesn't need an env variable, and `upstream/master` isn't always defined (my remotes are `alok` and `origin`). Setting this to `master` or `origin/master` is a better idea because `master` is meant to track the main branch.

File: .travis/yapf.sh [hunk 8]
```diff
@@ -9,17 +9,49 @@ builtin cd "$(dirname "${BASH_SOURCE:-$0}")"
...
+
+# Format specified files
+format() {
+    yapf $YAPF_FLAGS -- "$@"
+}
+
+# Format files that differ from main branch
+format_changed() {
+    FILES=`git diff --name-only $UPSTREAM_MASTER -- '*.py'`
+    yapf $YAPF_FLAGS $YAPF_EXCLUDES -- $FILES
```
Comment (alok) [inline .travis/yapf.sh]: This function should not exclude anything, since if you want that sort of fine-grained control over what you format, you can look at the output from `git status` manually and just pass those files to `format`. This is meant to be a quick way to format all files that differ from the main branch.

Comment (ericl) [reply .travis/yapf.sh]: If you want to do that, you can run with --files. Consider the common use case. Would a user be surprised if the auto-formatter formatted a file that is excluded?

**Comment by alok** [comment]
I don't think adding variables like that is a good idea. `format` is just meant to format the files you pass it with the minimum number of flags set to do it in parallel and to follow the style file. It should not exclude anything passed to it, that should be up to the user. `format_all` should use those flags, but it's kind of pointless if it's the only function using the exclude.

File: .travis/yapf.sh (hunk 7 above)
Comment (ericl) [reply .travis/yapf.sh]: Changed to `origin/master`. Note that you can override the target here by setting `RAY_UPSTREAM_BRANCH`.

**Comment by ericl** [comment]
An typical developer trying to use the yapf script isn't going to consider these things; they just want to run it and have it do the right thing, unconditionally. Having modes that don't obey excludes is just complicating the situation.

File: .travis/yapf.sh (hunk 7 above)
Comment (ericl) [reply .travis/yapf.sh]: Actually there's a better way to do this... WIP

Comment (ericl) [reply .travis/yapf.sh]: Ok... now we always fetch the latest upstream, and grab the merge base, so the diff should always be correct.

File: .travis/yapf.sh [hunk 9]
```diff
@@ -9,17 +9,54 @@ builtin cd "$(dirname "${BASH_SOURCE:-$0}")"
...
+YAPF_EXCLUDES="--exclude 'python/ray/cloudpickle' \
+--exclude 'python/ray/dataframe' \
+--exclude 'python/ray/rllib' \
+--exclude 'python/build' \
+--exclude 'python/ray/pyarrow_files' \
+--exclude 'python/ray/core/src/ray/gcs' \
+--exclude 'python/ray/common/thirdparty'"
+
+# Add the upstream branch if it doesn't exist
+if [ ! -e $ROOT/.git/refs/remotes/upstream ]; then
```
Comment (alok) [inline .travis/yapf.sh]: @ericl FYI, you should use `[[` in bash over `[`, since `[[` can warn of syntax errors because it's part of the syntax rather than an external command.

**Comment by robertnishihara** [comment]
Speaking of shell script best practices, @mehrdadn showed me https://github.com/koalaman/shellcheck recently. It integrates with various editors which is pretty nice.

File: scripts [hunk 10]
```diff
@@ -0,0 +1 @@
+.travis
```
Comment (robertnishihara) [inline scripts]: this was probably added by accident, right?

Comment (alok) [reply scripts]: No, this is on purpose. scripts is (currently) a symlink for .travis that can later be pulled out (without much difficulty) into a proper repo for one-off dev/user scripts. All of those scripts are in .travis right now, so I just symlinked it.

**Comment by alok** [comment]
@robertnishihara Yeah I use shellcheck for every line of shell I write, plus [shellharden](https://github.com/anordal/shellharden).

File: .travis/yapf.sh [hunk 11]
```diff
@@ -9,17 +9,49 @@ builtin cd "$(dirname "${BASH_SOURCE:-$0}")"
...
+UPSTREAM_MASTER=${RAY_UPSTREAM_BRANCH:-origin/master}
+
+# Format specified files
+format() {
+    yapf $YAPF_FLAGS -- "$@"
+}
+
+# Format files that differ from main branch
+format_changed() {
+    FILES=`git diff --name-only $UPSTREAM_MASTER -- '*.py'`
```
[the remark this answers is not in the document]
Comment (ericl) [reply .travis/yapf.sh]: Sounds good

File: .travis/yapf.sh [hunk 12]
```diff
@@ -7,19 +7,70 @@ set -eo pipefail
...
+format() {
+    yapf "${YAPF_FLAGS[@]}" -- "$@"
+}
+
+# Format files that differ from main branch. Ignores dirs that are not slated
+# for autoformat yet.
+format_changed() {
+    if ! git diff --diff-filter=ACM --quiet --exit-code 'upstream/master' 'HEAD' -- '*.py' &>dev/null; then
+        git diff --name-only 'upstream/master' 'HEAD' -- '*.py' | xargs -P 5 \
+            yapf "${YAPF_EXCLUDES[@]}" "${YAPF_FLAGS[@]}"
```
Comment (ericl) [inline .travis/yapf.sh]: Why do you need the xargs here when the yapf flag includes --parallel?

File: .travis/yapf.sh [hunk 13]
```diff
@@ -7,19 +7,70 @@ set -eo pipefail
...
+# Format specified files
+format() {
+    yapf "${YAPF_FLAGS[@]}" -- "$@"
+}
+
+# Format files that differ from main branch. Ignores dirs that are not slated
+# for autoformat yet.
+format_changed() {
+    if ! git diff --diff-filter=ACM --quiet --exit-code 'upstream/master' 'HEAD' -- '*.py' &>dev/null; then
+        git diff --name-only 'upstream/master' 'HEAD' -- '*.py' | xargs -P 5 \
```
Comment (ericl) [inline .travis/yapf.sh]: Shouldn't this be diffing against the mergebase? Have you checked what happens if master gets ahead of the current branch?

File: .travis/yapf.sh [hunk 14]
```diff
@@ -7,19 +7,70 @@ set -eo pipefail
...
+
+# Format specified files
+format() {
+    yapf "${YAPF_FLAGS[@]}" -- "$@"
+}
+
+# Format files that differ from main branch. Ignores dirs that are not slated
+# for autoformat yet.
+format_changed() {
+    if ! git diff --diff-filter=ACM --quiet --exit-code 'upstream/master' 'HEAD' -- '*.py' &>dev/null; then
```
Comment (ericl) [inline .travis/yapf.sh]: Let's avoid the double diff here. Just pipe the output directly into yapf.

File: .travis/yapf.sh [hunk 15]
```diff
@@ -7,19 +7,70 @@ set -eo pipefail
...
+        git diff --name-only 'upstream/master' 'HEAD' -- '*.py' | xargs -P 5 \
+            yapf "${YAPF_EXCLUDES[@]}" "${YAPF_FLAGS[@]}"
+    fi
+}
+
+# Formats *all* files that differ from main branch.
+format_all_changed() {
```
Comment (ericl) [inline .travis/yapf.sh]: This is unused. Please remove.

File: .travis/yapf.sh [hunk 16]
```diff
@@ -7,19 +7,70 @@ set -eo pipefail
...
+# This flag formats individual files. --files *must* be the first command line
+# arg to use this option.
+if [[ "$1" == '--files' ]]; then
+    format "${@:2}"
+    # If `--all` is passed, then any further arguments are ignored and the
+    # entire python directory is formatted.
+elif [[ "$1" == '--all' ]]; then
+    format_all
+else
+    # Format only the files that changed in last commit.
```
Comment (ericl) [inline .travis/yapf.sh]: changed from the upstream mergebase

**Comment by alok** [comment]
bump @ericl

**Review by ericl — CHANGES_REQUESTED** [review]

File: .travis/yapf.sh [hunk 17]
```diff
@@ -7,19 +7,72 @@ set -eo pipefail
...
+# Format files that differ from main branch. Ignores dirs that are not slated
+# for autoformat yet.
+format_changed() {
+    # The `if` guard ensures that the list of filenames is not empty, which
+    # could cause yapf to receive 0 positional arguments, making it hang
+    # waiting for STDIN.
+    #
+    # `diff-filter=ACM` is to ensure we only format files that exist on both
+    # branches.
+    if ! git diff --diff-filter=ACM --quiet --exit-code 'upstream/master' -- '*.py' &>dev/null; then
```
[the remark this answers is not in the document]
Comment (alok) [reply .travis/yapf.sh]: Oops, that's a typo.

File: .travis/yapf.sh [hunk 18]
```diff
@@ -7,19 +7,72 @@ set -eo pipefail
...
-    -- \
-    'test' 'python'
-
-if ! git diff --quiet; then
-    echo 'Reformatted staged files. Please review and stage the changes.'
+builtin cd "$ROOT" || exit 1
+
+# Add the upstream branch if it doesn't exist
+if ! [[ -e "$ROOT/.git/refs/remotes/upstream" ]]; then
+    git remote add 'upstream' 'https://github.com/ray-project/ray.git'
```
[the remark this answers is not in the document]
Comment (alok) [reply .travis/yapf.sh]: I put the single quotes to signify that those arguments are not part of the command `git remote add`, but that's a stylistic choice, so I'm happy to revert it.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Just tried this and it works now. Thanks for fixing it!

File: doc/source/development.rst [hunk 19]
```diff
@@ -54,7 +54,12 @@ helpful.
    something like ``flake8 ray/python/ray/worker.py``. You may need to first run
    ``pip install flake8``.
 
-5. **Inspecting Redis shards by hand:** To inspect the primary Redis shard by
+5. **Autoformatting code**. We use ``yapf`` https://github.com/google/yapf for
+   linting, and the config file is located at ``.style.yapf``. We recommend
+   running ``.travis/yapf.sh`` prior to pushing to format changed files.
```
[the remark this answers is not in the document]
Comment (alok) [reply doc/source/development.rst]: I think we should just hold off until we can look at all the scripts that belong in .travis and that belong in scripts. I made the symlink more for convenience, I don't expect it to stick around.
